### ISSUE TITLE
Issue with InvalidateBuiltInRoleCaches in db:migrate

### ISSUE THREAD

**Issue by gongomgra** [instructure/canvas-lms#1711]
Hello,

I'm having an issue running the `db:migrate` task for version `release/2020-09-23.32` with Ruby `ruby 2.5.8p224`. It seems to be related to the file `db/migrate/20200914184302_invalidate_built_in_role_caches.rb`, as if I rename this file to `20200914184302_invalidate_built_in_role_caches.rb.bck` and I run the `db:migrate` task again, it works fine

### Steps to reproduce:

`bundle exec ./bin/rake db:migrate RAILS_ENV=production` from the installation directory

### Expected behavior:

Database migration success.

### Actual behavior:

```
$ bundle exec ./bin/rake db:migrate RAILS_ENV=production
** Invoke db:migrate (first_time)
** Invoke db:load_config (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute db:load_config
** Execute db:migrate
1: production:public
==  InvalidateBuiltInRoleCaches: migrating ====================================
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

undefined method `global_id' for nil:NilClass
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/db/migrate/20200914184302_invalidate_built_in_role_caches.rb:43:in `up'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:817:in `exec_migration'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:798:in `block (2 levels) in migrate'
/home/admin/canvaslms-2020.09.23.32-0/ruby/lib/ruby/2.5.0/benchmark.rb:293:in `measure'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:797:in `block in migrate'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/switchman-1.15.2/lib/switchman/connection_pool_proxy.rb:13:in `with_connection'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:796:in `migrate'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/switchman-inst-jobs-1.5.2/lib/switchman_inst_jobs/active_record/migration.rb:16:in `migrate'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:608:in `migrate'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:977:in `migrate'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1292:in `block in execute_migration_in_transaction'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1345:in `ddl_transaction'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:1291:in `execute_migration_in_transaction'
/home/admin/canvaslms-2020.09.23.32-0/apps/canvaslms/htdocs/config/initializers/active_record.rb:1446:in `execute_migration_in_transaction'
...
```
### Additional notes:

After renaming the file to `20200914184302_invalidate_built_in_role_caches.rb.bck`, the command works fine

```
$ ls -l db/migrate/*invalidate*
-rw-r--r-- 1 admin admin 1931 Sep 30 01:11 db/migrate/20200914184302_invalidate_built_in_role_caches.rb.bck

$ bundle exec ./bin/rake db:migrate RAILS_ENV=production --trace
** Invoke db:migrate (first_time)
** Invoke db:load_config (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute db:load_config
** Execute db:migrate
1: production:public
==  BackfillDummyRootAccountIdOnFolders: migrating ============================
==  BackfillDummyRootAccountIdOnFolders: migrated (0.0928s) ===================

==  RegenerateBrandFilesBasedOnNewDefaultsPostdeploy: migrating ===============
==  RegenerateBrandFilesBasedOnNewDefaultsPostdeploy: migrated (0.1335s) ======

** Invoke db:_dump (first_time)
** Execute db:_dump

$ echo $?
0
```

**Comment by simonista** [comment]
That's a very strange error, it seems to indicate that you don't have a `Account.site_admin`, which is the administrative account. If you open a console (`./script/rails console production`) and try to query `Account.site_admin` is it indeed `nil`?

**Comment by gongomgra** [comment]
Hi @simonista,

Thanks for your reply. Yes, the command replied with `nil` for the query you sent. Originally I wasn't running the `db:migrate` command directly, but the `db:initial_setup` providing it with the admin credentials using environment variables like below, which I understand will take care of creating the required admin account

```
RAILS_ENV=production CANVAS_LMS_ADMIN_EMAIL="user@example.com" CANVAS_LMS_ADMIN_PASSWORD="mypassword" CANVAS_LMS_ACCOUNT_NAME="UserName LastName" CANVAS_LMS_STATS_COLLECTION=anonymized bundle exec rake db:initial_setup
```

Should I run the `db:configure_admin` task before calling `db:initial_setup` instead?

**Comment by danielnewell** [comment]
I'm having the same issue, @gongomgra. Attempting `db:configure_admin` causes `pseudonyms` to be undefined. Your command you gave gives the original issue.

@simonista any idea on how to fix?

**Comment by sanjibnarzary** [comment]
I have the same problem while trying to install the latest master version on Ubuntu 18.04 LTS Server in production mode

```
-----other logs

1: production:public
==  InvalidateBuiltInRoleCaches: migrating ====================================
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

undefined method `global_id' for nil:NilClass
/opt/canvas/db/migrate/20200914184302_invalidate_built_in_role_caches.rb:43:in `up'
```

**Comment by lferf** [comment]
I also have the same problem that @sanjibnarzary reports. I'm trying to install it on Ubuntu 16 LTS Server in production mode. 

`==  InvalidateBuiltInRoleCaches: migrating ====================================
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

undefined method `global_id' for nil:NilClass
/var/canvas/db/migrate/20200914184302_invalidate_built_in_role_caches.rb:43:in `up'
/var/canvas/vendor/bundle/ruby/2.6.0/gems/activerecord-5.2.4.4/lib/active_record/migration.rb:817:in `exec_migration'`

**Comment by gongomgra** [comment]
Hi @simonista, can you check this when you have a chance?

**Comment by you-mu** [comment]
I had the same problem on debian 10.6.

 ```
constant ::Fixnum is deprecated
1: production:public
1: production:public
==  InvalidateBuiltInRoleCaches: migrating ====================================
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

undefined method `global_id' for nil:NilClass
```

**Comment by simonista** [comment]
Looks like this got patched here: 404bb682c1dcfcc65252fe3a65dc42c66e11d6c0, although I still don't understand why this would be happening :(. But feel free to cherry-pick that onto your branch if it's happening to you.

**Comment by gongomgra** [comment]
Hi @simonista,

Thanks for linking the patch. I see there is a new release including that minor change (`release/2020-10-21.15`), but I'm having another issue now. Could you check it?

```
==  InvalidateBuiltInRoleCaches: migrating ====================================
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

undefined method `clear_downstream_caches' for nil
/home/admin/canvaslms-2020.10.21.15-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/inst-jobs-0.16.0/lib/delayed/performable_method.rb:4:in `initialize'
/home/admin/canvaslms-2020.10.21.15-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/inst-jobs-0.16.0/lib/delayed/message_sending.rb:35:in `new'
/home/admin/canvaslms-2020.10.21.15-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/inst-jobs-0.16.0/lib/delayed/message_sending.rb:35:in `send_later_enqueue_args'
/home/admin/canvaslms-2020.10.21.15-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/switchman-inst-jobs-2.0.0/lib/switchman_inst_jobs/delayed/message_sending.rb:7:in `send_later_enqueue_args'
/home/admin/canvaslms-2020.10.21.15-0/apps/canvaslms/htdocs/vendor/bundle/ruby/2.5.0/gems/inst-jobs-0.16.0/lib/delayed/message_sending.rb:75:in `send_later_if_production_enqueue_args'
/home/admin/canvaslms-2020.10.21.15-0/apps/canvaslms/htdocs/db/migrate/20200914184302_invalidate_built_in_role_caches.rb:40:in `up'
(...)
```

**Comment by simonista** [comment]
Hi @gongomgra, that looks like the same basic issue, but coming from not running production mode. Possible you aren't running with `RAILS_ENV=production` ?

**Comment by gongomgra** [comment]
Hi @simonista,

Thanks, I'm ensuring the `RAILS_ENV=production` is present

```
RAILS_ENV=production CANVAS_LMS_ADMIN_EMAIL="user@example.com" CANVAS_LMS_ADMIN_PASSWORD="_REDACTED_" CANVAS_LMS_ACCOUNT_NAME="UserName LastName" CANVAS_LMS_STATS_COLLECTION=anonymized bundle exec rake db:initial_setup
```

I also tried running an `export RAILS_ENV=production` before running the rake task, but to no avail.

**Comment by simonista** [comment]
Okay, I misread the stack trace. I think I see the problem. When you use `db:initial_setup`, it runs post deploy migrations before forcing creation of the site admin account. In non-production mode, this doesn't matter, the account is just created the first time it's asked for, but in prod that doesn't happen. I'll get a commit up to fix this, but for now you could change line 38 to `if Shard.current.default? && Account.site_admin`

**Comment by gongomgra** [comment]
Thank you @simonista,

It worked. Do you plan to include that change in a future release?

**Comment by simonista** [comment]
This is fixed now in stable here: 361391a4794e4bbdb9bee3895ade6695cab537e2
